### DISCUSSION TITLE
Error occuring  during the running of cml file

### DISCUSSION THREAD

**Discussion by Alimiji** [community/community#57478] [Programming Help]
### Hi everyone,

Hi everyone, 
I need your help please. This is the error that occurs: 
===================================================================================================
{"code":"ERR_INVALID_URL","input":"<html>\r\n<head><title>400 Bad Request</title></head>\r\n<body>\r\n<center><h1>400 Bad Request</h1></center>\r\n<hr><center>cloudflare</center>\r\n</body>\r\n</html>\r\n","level":"error","message":"Invalid URL","stack":"TypeError [ERR_INVALID_URL]: Invalid URL\n    at new NodeError (node:internal/errors:399:5)\n    at new URL (node:internal/url:560:13)\n    at uriParam (/usr/local/lib/node_modules/@dvcorg/cml/src/utils.js:151:15)\n    at watermarkUri (/usr/local/lib/node_modules/@dvcorg/cml/src/utils.js:142:10)\n    at CML.publish (/usr/local/lib/node_modules/@dvcorg/cml/src/cml.js:330:13)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async visitor (/usr/local/lib/node_modules/@dvcorg/cml/src/cml.js:229:24)\n    at async Promise.all (index 0)\n    at async publishLocalFiles (/usr/local/lib/node_modules/@dvcorg/cml/src/cml.js:242:7)"}
Error: Process completed with exit code 1.
====================================================================================================
My CML code is:

name: CML & DVC
on: [push]
jobs:  
  train-and-report:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-python@v4
        with:
          python-version: '3.8.16'
      - uses: iterative/setup-cml@v1
      - uses: iterative/setup-dvc@v1
      - name: Train model
        run: |
          pip install -r requirements.txt
          python model.py
      - name: Create CML report
        env:
          REPO_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: |
          echo '## Data Summary' >> report.md
          python -c "import pandas as pd; df = pd.read_csv('Data/Incidents_metro.csv'); print(df.head(10))" >> report.md
          echo '## Column Histogram' >> report.md
          python -c "import pandas as pd; import matplotlib.pyplot as plt; df = pd.read_csv('Data/Incidents_metro.csv'); df.hist(figsize=(10, 10)); plt.savefig('histogram.png')"
          if [ -f histogram.png ]; then
            echo 'Histogram image created successfully.'
          else
            echo 'Histogram image creation failed.'
            exit 1
          fi
          echo '![Histogram](histogram.png)' >> report.md
          cml comment create report.md

**Comment by Pipe-Vash** [comment]
I am having a similar problem. In my case is:

`{"code":"ERR_INVALID_URL","input":"<html>\r\n<head><title>400 Bad Request</title></head>\r\n<body>\r\n<center><h1>400 Bad Request</h1></center>\r\n<hr><center>cloudflare</center>\r\n</body>\r\n</html>\r\n","level":"error","message":"Invalid URL","stack":"TypeError [ERR_INVALID_URL]: Invalid URL\n    at new NodeError (node:internal/errors:399:5)\n    at new URL (node:internal/url:560:13)\n    at uriParam (/usr/local/lib/node_modules/@dvcorg/cml/src/utils.js:149:15)\n    at watermarkUri (/usr/local/lib/node_modules/@dvcorg/cml/src/utils.js:140:10)\n    at CML.publish (/usr/local/lib/node_modules/@dvcorg/cml/src/cml.js:349:13)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async visitor (/usr/local/lib/node_modules/@dvcorg/cml/src/cml.js:247:26)\n    at async Promise.all (index 0)\n    at async publishLocalFiles (/usr/local/lib/node_modules/@dvcorg/cml/src/cml.js:261:7)"}
Error: Process completed with exit code 1.`

While my .yaml is the following:
`
name: Continuous Training
on:
  push:
    branches:
    - workflow_continuous_training
  schedule:
    - cron: '0 */6 * * *'
  workflow_dispatch:
    inputs:
      reason:
        description: Why to run this?
        required: false
        default: running CT
jobs:
  continuous-training:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout Repo
        uses: actions/checkout@v2
      - uses: iterative/setup-dvc@v1
      - name: Set up Python 3.8
        uses: actions/setup-python@v2
        with:
          python-version: 3.8

      - name: Train model
        env:
          SERVICE_ACCOUNT_KEY: ${{ secrets.SERVICE_ACCOUNT_KEY }}
        run: |
          pip3 install virtualenv
          virtualenv venv
          source venv/bin/activate
          pip install -r src/requirements.txt
          export GOOGLE_APPLICATION_CREDENTIALS=$(python utilities/setter.py)
          dvc pull model/model.pkl.dvc -r model-track
          dvc unprotect model/model.pkl
          dvc repro -f 
          echo "Training Completed"
          dvc add model/model.pkl -r model-track --to-remote
          dvc push model/model.pkl.dvc -r model-track
      - name: Commit .dvc file changes
        run: |
          git config --local user.email "pipe.vash@gmail.com"
          git config --local user.name "github-actions[bot]"
          git add model/model.pkl.dvc
          git commit -m "Updating model serialization"
      - uses: ad-m/github-push-action@master
        with:
            github_token: ${{ secrets.GITHUB_TOKEN }}
            branch: ${{ github.ref }}
      - uses: iterative/setup-cml@v1
      - name: Push metrics
        env:
          REPO_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: |
          cat report.txt >> report.md
          echo '![](./prediction_behavior.png "Prediction Behavior")' >> report.md
          cml comment create report.md
`

I have tried with different methods, but none of them works. Any help would be welcomed

**Comment by hwang9u** [comment]
@Alimiji @Pipe-Vash 

Hi :) 
I had same problem and I resolved it!
First, I added node version (I think this didn't solve the error directly)
**Second, I made new github token and named "TOKEN".**
Because Github told me that "GITHUB_" should not be included in the token name.
So I changed **secrets.GITHUB_TOKEN -> secrets.TOKEN** and resolved it finally!
I hope this helps you!!!

<img width="551" alt="image" src="https://github.com/community/community/assets/48237908/48995caf-275c-412b-8f3d-4b06273f2528">

<img width="552" alt="image" src="https://github.com/community/community/assets/48237908/92cf13e3-864a-4238-b80b-64273f51ec93">

**Reply by luyangliuable** [reply]
For the TOKEN did you put down your PAT as your secret?

**Reply by GuillermoSainz07** [reply]
thanks, this really solved my problem!!
![image](https://github.com/user-attachments/assets/b510e9c4-54e1-41f1-bdd6-66e83faf8ced)
![image](https://github.com/user-attachments/assets/c64c0ba5-8f4e-4dab-9c1e-d6f511d546dd)

**Comment by samus-aran** [comment]
Welcome to the GitHub Community, @Alimiji, we're happy you're here! You are more likely to get a useful response if you are posting your question(s) in the applicable category and are explicit about what your project entails--giving a few more details might help someone give you a nudge in the right direction. I've gone ahead and moved it for you. Good luck!e

**Comment by rushabhcodes** [comment]
A "400 Bad Request" error is an HTTP status code that indicates the server could not understand the request sent by the client, often due to malformed or invalid syntax. This error typically occurs when there's an issue with the request sent to a web server, such as missing or incorrect parameters, headers, or data.

Here are some common reasons and solutions for encountering a "400 Bad Request" error:

1. Malformed URL: Ensure that the URL you are accessing is properly formatted, and there are no typos or special characters that could be causing issues.

2. Missing or invalid request parameters: Check if the required parameters for the request are present and have the correct format. If there are optional parameters, make sure they are provided correctly.

3. Invalid headers: Verify that any custom headers you're using are spelled correctly and are in the correct format.

4. Request content issue: If you're sending data with the request (e.g., in the case of a POST request), make sure the data is properly formatted and matches the expected structure.

5. Request size limit: Some servers may impose size limits on incoming requests. If your request is too large, it could result in a "400 Bad Request" error. Check if there are any size restrictions and try reducing the request size if necessary.

6. Cookies and caching: Sometimes, issues with cookies or cached data can lead to this error. Clear your browser's cache and cookies and try again.

7. Browser extensions: Certain browser extensions or plugins may interfere with the request, causing a "400 Bad Request" error. Try accessing the URL in an incognito/private browsing window to rule out any extension-related issues.

8. Server-side issues: Although less likely, the problem could also be on the server-side. The server might not be configured correctly to handle the request, or there could be a bug in the server code.

If you are the developer of the web application, you can check the server logs for more detailed information about the specific reason for the "400 Bad Request" error. Examining the logs can help you pinpoint the issue and fix it accordingly.

If the error persists and you are not in control of the server or application, you may want to reach out to the website's support team or the developers responsible for maintaining the service to report the issue and seek assistance.

**Comment by SAMIULLAH-KALHORO-DEV** [comment]
The error message you provided indicates that there was an issue with the URL or the request made to the server. The server responded with a "400 Bad Request" status code, indicating that the request was malformed or invalid in some way.

This error can occur for various reasons, such as:

1. Invalid URL: The URL used to make the request might be incorrect or improperly formatted.

2. Missing Required Data: The request may be missing some required data or parameters.

3. Invalid HTTP Method: The HTTP method used in the request (GET, POST, etc.) might not be supported or appropriate for the particular URL.

4. Authentication/Authorization Issues: The server may require authentication, and the request does not include the necessary credentials.

5. Server-Side Issue: There might be an issue on the server side that is causing the error.

To resolve this issue, you can try the following steps:

1. Double-check the URL: Make sure the URL you are trying to access is correct and properly formatted. Check for any typos or missing parts.

2. Check Request Parameters: If your request includes parameters or data, ensure that they are provided correctly.

3. Use the Appropriate HTTP Method: Verify that you are using the correct HTTP method (GET, POST, etc.) for the specific endpoint.

4. Check for Authentication/Authorization: If the server requires authentication, make sure to include the necessary credentials in your request.

5. Retry the Request: Sometimes, a temporary issue on the server can cause a "400 Bad Request" error. Retry the request after some time to see if the problem persists.

If you continue to encounter the error, there may be specific details in the server's response that could provide more insights into the issue. Checking the server logs or contacting the server administrators might be helpful in diagnosing the problem further.

**Comment by ajsalkp** [comment]
me also facing similar issue with cml . anybody can help on this issue?
![image](https://github.com/community/community/assets/34513997/eecae1c0-0006-4a54-91b3-8f1b90cc4b1e)
